tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java

```diff
@@ -0,0 +1,697 @@
+
+
+   /*
+    use case is dispatch from memory, with non-blocking producers
+    producers add to the head of the broker chain, consumers receive from the tail
+    when head == tail we are back to one broker for that address, the end of the chain
+   */
+   private void prepareNodesAndStartCombinedHeadTail() throws Exception {
+      prepareNodesAndStartCombinedHeadTail(1000, 300);
+   }
+
+   private void prepareNodesAndStartCombinedHeadTail(int credit, int creditMin) throws Exception {
+      Assert.assertTrue(credit > 0);
+
+      AddressSettings blockingQueue = new AddressSettings();
+      blockingQueue.setMaxSizeBytes(100 * 1024)
+         .setAddressFullMessagePolicy(AddressFullMessagePolicy.FAIL)
+         .setSlowConsumerPolicy(SlowConsumerPolicy.KILL).setSlowConsumerThreshold(0).setSlowConsumerCheckPeriod(TimeUnit.MILLISECONDS.toSeconds(opTimeoutMillis))
+         .setAutoDeleteQueues(false).setAutoDeleteAddresses(false); // so slow consumer can kick in!
+
+      Configuration baseConfig = new ConfigurationImpl();
+      baseConfig.getAddressesSettings().put(qName, blockingQueue);
+
+      BrokerBalancerConfiguration balancerConfiguration = new BrokerBalancerConfiguration();
+      balancerConfiguration.setName(balancerConfigName).setTargetKey(TargetKey.ROLE_NAME).setTargetKeyFilter("(?<=^EQ_).*"); // strip EQ_ prefix
+      baseConfig.addBalancerConfiguration(balancerConfiguration);
+
+      // prepare two nodes
+      for (int nodeId = 0; nodeId < 2; nodeId++) {
+         Configuration configuration = baseConfig.copy();
+         configuration.setName("Node" + nodeId);
+         configuration.setBrokerInstance(new File(getTestDirfile(), configuration.getName()));
+         configuration.addAcceptorConfiguration("tcp", "tcp://localhost:" + (base_port + (nodeId)) + "?redirect-to=" + balancerConfigName + ";amqpCredits=" + credit + ";amqpMinCredits=" + creditMin);
+         nodes.add(new EmbeddedActiveMQ().setConfiguration(configuration));
+         nodes.get(nodeId).setSecurityManager(customSecurityManager);
+         nodes.get(nodeId).setMbeanServer(mBeanServer);
+      }
+
+      // node0 initially handles both producer & consumer (head & tail)
+      nodes.get(0).getConfiguration().getBalancerConfigurations().get(0).setLocalTargetFilter("PRODUCER|CONSUMER");
+      nodes.get(0).start();
+   }
+
+   @Test (timeout = 20000)
+   public void testScale0_1() throws Exception {
+
+      prepareNodesAndStartCombinedHeadTail();
+
+      // slow consumer, delay on each message received
+      EQConsumer eqConsumer = new EQConsumer(urlForNodes(nodes));
+      eqConsumer.start();
+
+      // verify consumer reconnects on no messages
+      assertTrue(Wait.waitFor(() -> eqConsumer.connectionCount.get() > 1));
+
+      EQProducer eqProducer = new EQProducer(urlForNodes(nodes));
+      eqProducer.start();
+
+      // verify producer reconnects on fail full!
+      assertTrue(Wait.waitFor(() -> eqProducer.connectionCount.get() > 1));
```

Comment:
why not using Wait.assertTrue ?

Reply:
Also the test has an overall 20sec timeout, but this single check (and all the other ones) has a 30sec timeout. Seems like the checks should be gated by how long they are reasonably expected to take.